### PR TITLE
prop-types package definitions

### PR DESCRIPTION
**Pull request by DovydasNavickas**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not provide its own types, and you can not add them.
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, and `strictNullChecks` set to `true`.

### CODE REVIEW

**Comment by DovydasNavickas** [comment]
Hmm.... In the build report here:
https://travis-ci.org/DefinitelyTyped/DefinitelyTyped/builds/220110357

I get an error:
```
Error in prop-types
Command failed: node /home/travis/build/DefinitelyTyped/DefinitelyTyped/node_modules/types-publisher/bin/tester/../../node_modules/dtslint/bin/index.js --dt
/home/travis/build/DefinitelyTyped/DefinitelyTyped/types/prop-types/prop-types-tests.ts
ERROR: 1:28  expect  TypeScript compile error: File '/home/travis/build/DefinitelyTyped/DefinitelyTyped/types/prop-types/index.d.ts' is not a module.
```

The essential part is:
`File '/home/travis/build/DefinitelyTyped/DefinitelyTyped/types/prop-types/index.d.ts' is not a module.`

I tried declaring a namespace and exporting it:
```ts
declare namespace PropTypes {
    //<...>
}

export = PropTypes;
```

But it said not to use `export =`. I'm a bit confused, how should I export everything here?

**Comment by DovydasNavickas** [comment]
Ooook... Now I'm super confused or my googling skills are insufficient.

Now I get this:
```no-single-declare-module  File has only 1 module declaration — write it as an external module.```

And I cannot find what the "external module", or at least a *newest* definition of it that makes sense.

@andy-ms, @RyanCavanaugh, could you explain that briefly?

**Comment by andy-ms** [comment]
"external module" just means writing exports at the top level in standard ES6 style. See `base-64/index.d.ts` for an example.

File: types/prop-types/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+    interface Requireable<T> extends Validator<T> {
+        isRequired: Validator<T>;
+    }
+
+    type ValidationMap<T> = {[K in keyof T]?: Validator<T> };
+
+    let any: Requireable<any>;
+    let array: Requireable<any>;
+    let bool: Requireable<any>;
```
Comment (andy-ms) [inline types/prop-types/index.d.ts]: Should this be `Requireable<boolean>`?
Also, prefer `const` if this is not intended to be modified.

File: types/prop-types/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    let func: Requireable<any>;
+    let number: Requireable<any>;
+    let object: Requireable<any>;
+    let string: Requireable<any>;
+    let node: Requireable<any>;
+    let element: Requireable<any>;
+    function instanceOf(expectedClass: {}): Requireable<any>;
+    function oneOf(types: any[]): Requireable<any>;
+    function oneOfType(types: Array<Validator<any>>): Requireable<any>;
+    function arrayOf(type: Validator<any>): Requireable<any>;
```
Comment (andy-ms) [inline types/prop-types/index.d.ts]: Is there a relation between the input and output types? Maybe something like `arrayOf<T>(type: Validator<T>): Requireable<T[]>`?

File: types/prop-types/prop-types-tests.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import * as PropTypes from "prop-types";
+
+let testObj: any = {
```
Comment (andy-ms) [inline types/prop-types/prop-types-tests.ts]: This doesn't really test out the library. All you've done is put everything in a bag of type `any`. If the types were wrong, this code would still work. Try including realistic uses of the provided functions and objects.

**Comment by DovydasNavickas** [comment]
@andy-ms, I tried to declare everything instead of exporting. Thanks for the heads up.

Now the build is passing.

Testing this library is a bit tricky, though, because its values are being used in a static properties of `React.Component<TProps, TState>` (and static properties are not included in d.ts) or in properties of type `ValidationMap<any>`.

And [`ValidationMap<T>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L216-L218) is a dictionary of Validators:
```ts
type ValidationMap<T> = { [K in keyof T]?: Validator<T> };
```

I've updated the test and I guess it is pretty close to realistic use-case now.

**Comment by DovydasNavickas** [comment]
An answer to your comments:
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/15736/files/92633c2e1ddbf2f178ccbf7d83f397c96bfb9caf#r110530373
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/15736/files/92633c2e1ddbf2f178ccbf7d83f397c96bfb9caf#r110530390

Specifying `T` would be an incorrect type check here. The package is for checking the propTypes during runtime, not compile time.
And types do not flow to runtime, unfortunately.

And as you can see here:
https://github.com/reactjs/prop-types/blob/master/factory.js#L75-L84

The `T` given in `Requireable<T>` is the `object: T`, which is `props` in React, not the type you require prop value to be:
```ts
import * as React from "react";
import * as PropTypes from "prop-types";

interface MyProps {
    name: string;
}

class MyComponent extends React.Component<MyProps, {}> {
    static propTypes: PropTypes.ValidationMap<MyProps> = {
        name: PropTypes.string.isRequired
    }
}
```

And this is why we cannot put a specific type into `Requireable<T>` there.

**Comment by n1ghtmare** [comment]
Anyway we can get this merged?

**Pull request merged**